src/main/scala/metadata.scala

```diff
@@ -0,0 +1,83 @@
+package suffuse
+
+/** AKey is an attribute key: a Metadata map key which carries
+ *  both the type of the corresponding value, and a value to
+ *  provide when there is no value in the map. This as opposed
+ *  to wrapping everything in sight in Option.
+ */
+final class AKey[A](description: String, empty: => A) {
+  def emptyValue: A = empty
+  override def toString = description
+}
+
+/** Attr is a dependently typed key/value pair.
+ */
+trait Attr extends Any {
+  type Type
+  def key: AKey[Type]
+  def value: Type
+
+  def pair                               = key -> value
+  def matches[A](implicit that: AKey[A]) = key == that
+  def to_s                               = s"$key: $value"
+}
+
+/** The simple implementation of Attr, which ties the knot between the
+ *  user-facing type parameter and the type member which couples key and value.
+ */
+final case class AttrOf[A](key: AKey[A], value: A) extends Attr {
+  type Type = A
+  override def toString = s"$key: $value"
+}
+
+object Attr {
+  def apply[A](value: A)(implicit key: AKey[A]): AttrOf[A] = new AttrOf[A](key, value)
+  def unapply(x: Attr): Some[(AKey[x.Type], x.Type)]       = Some(x.pair)
+}
+
+/** A Metadata map, holding any number of well typed Attrs.
+ *  A typed value can be obtained for any key.
+ */
+final class Metadata(attrs: Vector[Attr]) {
+  private[this] val untypedMap                   = attrs.foldLeft(Map[AKey[_], Any]())(_ + _.pair)
+  private def doApply[A]()(implicit z: AKey[A]): A = untypedMap(z).asInstanceOf[A]
+
+  def isEmpty                                = untypedMap.isEmpty
+  def apply[A]()(implicit z: AKey[A]): A     = if (has[A]) doApply[A]() else z.emptyValue
+  def has[A]()(implicit z: AKey[A]): Boolean = keys contains z
+  def keys: Vector[AKey[_]]                  = attrs map (_.key)
+
+  /** Set could also hold onto the old value, and just return the last one.
+   *  It would amount to preserving the attribute's history.
+   */
+  def set(attr: Attr): Metadata        = new Metadata(attrs.filterNot(_.key eq attr.key) :+ attr)
+  def set[A: AKey](value: A): Metadata = set(Attr[A](value))
+
+  override def toString = if (isEmpty) "{ }" else attrs mkString ("{\n  ", "\n  ", "\n}")
+}
+object Metadata {
+  private[this] val Empty: Metadata = new Metadata(Vector())
+
+  def empty(): Metadata             = Empty
```

Comment:
It's the other way around. A def is (if written as such) referentially transparent, but a val cannot be because its location in memory can be inspected. And because a val has to be placed into a memory location, it will always side effect the memory store, which creates exactly the usual problems with composing side effects - except in this case we call those problems "initialization order issues", often seen when something turns up null.

Reference equality is equivalent to the runtime exposing the map from object references to memory locations. When you ask whether `a eq b` it's equivalent to looking up both references in the map and seeing if they have the same value. This significantly alters the nature of equality. A functional value is fully defined by its elements. As soon as you introduce reference equality, you cannot be fully defined by your elements. It's why you see behavior like this:
```
scala> "" eq ""
res0: Boolean = true

scala> "" eq new String("")
       "" eq new String("")
          ^
res1: Boolean = false

scala> "" == new String("")
res2: Boolean = true
```
Are two empty strings equal? As long as we can't see into that big memory location map, we have no means of distinguishing two empty strings. They are definitionally equal. But if we can, then every empty string carries its own behavior.